### PR TITLE
[Docs] AIR batch forecasting Prophet example (SSML)

### PR DESCRIPTION
**Pull request by christy**
## Why are these changes needed?

This example follows the same AIR pattern as [Batch training & tuning on Ray Tune](https://github.com/ray-project/ray/blob/master/doc/source/ray-air/examples/batch_tuning.ipynb).  Also used the same NYC Taxi data.  However we use Prophet time series forecasting models instead of Scikit-learn regression models.

Things still left to do IMHO:
- Still need to polish notebook explanatory texts.
- I noticed some inconsistency restoring my inference pandas df from checkpoint.  (Prophet outputs a nice pandas df that includes confidence intervals).  I will double check my code.  I often had to re-inference the restored-from-checkpoint model again in order to see a correct plot at end of notebook.
- The example of tuning a seasonality hyperparameter in Prophet might not be that useful?  Since Prophet is good at automatically trying and using the right seasonalities.  
- Maybe a better tuning example would be comparing 2 different Prophet algorithm flavors? Original [Prophet](https://github.com/facebook/prophet) vs research team maintained [Prophet] (https://github.com/facebookresearch/Kats/blob/main/kats/models/prophet.py) and let both algorithms do their thing to automatically detect the right seasonal effects.  Another option is ARIMA vs Prophet as a Tune experiment.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
I'd try both flavors of prophet and autoarmia from statsforecast.

**Comment by christy** [comment]
Latest commit:
- Updated notebook explanatory texts.
- Fixed restoring Prophet forecast pandas dataframes from checkpoint, now I see correct plots at end of notebook.
- Replaced Tune experiment so it uses 2 flavors of Prophet seasonality + ARIMA
- One more example that might be useful is replace the PyArrow data step with Ray Datasets

**Review by Yard1 — APPROVED** [review]
LGTM, thanks!

File: doc/source/ray-air/examples/batch_forecasting.ipynb [hunk 1]
```diff
@@ -0,0 +1,1444 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c2953d85",
+   "metadata": {},
+   "source": [
+    "In this notebook, we will use Ray Tune to run parallel training jobs per pickup location. The training jobs will be defined using a search space and simple grid search. Depending on your need, fancier search spaces and search algorithms are possible with Tune.\n",
+    "\n",
+    "**First, define a search space of experiment trials to run.**\n",
+    "> The typical use case for Tune search spaces are for hypterparameter tuning. In our case, we are defining a Tune search space in a way to allow for training jobs to be conducted automatically. Each training job will run on a different data partition (taxi pickup location) and use a different model.\n",
```
Comment (richardliaw) [inline doc/source/ray-air/examples/batch_forecasting.ipynb:440]:
```suggestion
    "> The typical use case for Tune search spaces are for hyperparameter tuning. In our case, we are defining a Tune search space in a way to allow for training jobs to be conducted automatically. Each training job will run on a different data partition (taxi pickup location) and use a different model.\n",
```

File: doc/source/ray-air/examples/batch_forecasting.ipynb [hunk 2]
```diff
@@ -0,0 +1,1444 @@
...
+    "\n",
+    "**Next, define a search algorithm.**\n",
+    ">Ray Tune will use the search space and the specified search algorithm to generate multiple configurations, each of which will be evaluated in a separate Trial on a Ray Cluster. Ray Tune will take care of orchestrating those Trials automatically. Specifically, Ray Tune will pass a config dictionary to each partition and make a Trainable function call.\n",
+    "\n",
+    "**Below, we define our search space consists of:**\n",
+    "- Prophet with [multiplicative or additive](https://facebook.github.io/prophet/docs/multiplicative_seasonality.html) seasonal effects or AutoARIMA,\n",
+    "- Some or all NYC taxi drop-off locations.\n",
+    "\n",
+    "**Also below, we define our search algorithm as:**\n",
+    "- Grid search.\n",
```
Comment (richardliaw) [inline doc/source/ray-air/examples/batch_forecasting.ipynb:452]: this seems a bit weird

File: doc/source/ray-air/examples/batch_forecasting.ipynb [hunk 3]
```diff
@@ -0,0 +1,1444 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "642d21d4",
+   "metadata": {},
+   "source": [
+    "**In the cell below, we configure the resources allocated per trial.** \n",
+    "\n",
+    "Tune uses this resources allocation to control the parallelism. For example, if each trial was configured to use 4 CPUs, and the cluster had only 32 CPUs, then Tune will limit the number of concurrent trials to 8 to avoid overloading the cluster. For more information, see [A Guide To Parallelism and Resources](https://docs.ray.io/en/master/tune/tutorials/tune-resources.html#tune-parallelism)."
+   ]
```
Comment (richardliaw) [inline doc/source/ray-air/examples/batch_forecasting.ipynb:627]: We can skip this section (or put it in the bottom); it's not critical path for the user to know

File: doc/source/ray-air/examples/batch_forecasting.ipynb [hunk 4]
```diff
@@ -0,0 +1,1444 @@
...
+    }
+   ],
+   "source": [
+    "# get a list of training loss errors\n",
+    "errors = []\n",
+    "[errors.append(i.metrics.get(\"error\", 10000.0)) for i in results]\n",
+    "\n",
+    "# get a list of checkpoints\n",
+    "checkpoints = []\n",
+    "[checkpoints.append(i.checkpoint) for i in results]\n",
```
Comment (richardliaw) [inline doc/source/ray-air/examples/batch_forecasting.ipynb:998]:
```suggestion
    "checkpoints = [i.checkpoint for i in results]\n",
```

File: doc/source/ray-air/examples/batch_forecasting.ipynb [hunk 5]
```diff
@@ -0,0 +1,1444 @@
...
+    "\n",
+    "# get a list of locations\n",
+    "locations = []\n",
+    "[locations.append(i.config[\"location\"]) for i in results]\n",
+    "\n",
+    "# get a list of model params\n",
+    "model_type = []\n",
+    "[model_type.append(i.config[\"model_type\"]) for i in results]\n",
```
Comment (richardliaw) [inline doc/source/ray-air/examples/batch_forecasting.ipynb:1006]: let's make this more idiomatic (see suggested change above)

**Review by richardliaw — CHANGES_REQUESTED** [review]
Awesome work! some changes requested. Can we rename this to something like "parallel demand forecasting"?

**Comment by christy** [comment]
I'll pull again, and re-work those sections.

**Comment by richardliaw** [comment]
@christy the changes haven't been fixed yet; can you take a look?

**Comment by christy** [comment]
Made changes for comments https://github.com/ray-project/ray/pull/30430#pullrequestreview-1189141567

One thing that had me puzzled for a while, `train_test_split()` was switched out to scikit-learn's method. That does not work on time series. So another change in this latest commit was to sort the data by date, then manually split data into train, test.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**